tracer/src/Datadog.Trace/AppSec/ObjectExtractor.cs

```diff
@@ -51,44 +56,94 @@ private static bool IsOurKindOfPrimitive(Type t)
 
             visited.Add(body);
 
-            Log.Debug("ExtractProperties - body: {Body}", body);
+            if (Log.IsEnabled(LogEventLevel.Debug))
+            {
+                Log.Debug("ExtractProperties - body: {Body}", body);
+            }
 
-            var fields = body.GetType()
-                        .GetFields(BindingFlags.Instance | BindingFlags.NonPublic)
-                        .Where(x => x.IsPrivate && x.Name.EndsWith("__BackingField"))
-                        .ToArray();
+            var bodyType = body.GetType();
 
             depth++;
 
-            var dictSize = Math.Min(WafConstants.MaxContainerSize, fields.Length);
+            FieldExtractor[] fieldExtractors = null;
+            lock (Sync)
+            {
+                if (!TypeToExtractorMap.TryGetValue(bodyType, out fieldExtractors))
```

Comment:
If I understand correctly, this will run for every request, which means a lot of potential concurrency. Have you considered dropping the lock, and using a `ConcurrentDictionary<>` instead? Worst case, it's a bit of duplicate work for new bodyTypes, but that shouldn't slow, and should reduce contention in the default case where the body type has been seen before? I suspect it wouldn't show up in the benchmarks, unless you add one that does concurrent extraction...

Reply:
My guess was that we wanted to avoid duplicating the work of generating the dynamic method as it's expensive because of the jit compilation phase. Though I guess the only way to know is to do the experiments.

Reply:
I agree in general, but currently this blocks _all_ requests whenever _any_ new body is being generated. And it blocks them for the duration that it takes to do the generation. That essentially reduces allocations by ensuring there's no re-work, but at the expense of worse throughput. 

e.g. imagine the following 4 requests come in concurrently for types `TypeA` and `TypeB`. `TypeA` has already been generated. 

With the Lock:
- `TypeB`: takes the lock and starts generating, executes on completion
- `TypeB`: is blocked until the first `TypeB` is generated. Then executes.
- `TypeA`: is blocked until the first `TypeB` is generated. Then executes (with lock contention).
- `TypeA`: is blocked until the first `TypeB` is generated. Then executes (with lock contention).

With the Dictionary:
- `TypeB`: Starts generating, executes on completion.
- `TypeB`: Starts generating, executes on completion.
- `TypeA`: Not blocked, executes immediately
- `TypeA`: Not blocked, executes immediately

Reply:
Interesting, thanks for taking a look 👍